Aligning Wallet Instance Attestation with Attestation-Based Client Authentication via HTTP Headers (#327)
italia/eidas-it-wallet-docs: docs/en/pid-eaa-issuance.rst

        - National eID Relying Party: The component to authenticate the User with the national Digital Identity Providers, based on OpenID Connect Core 1.0 or SAML2.
        - National Identity Provider: It represents preexisting identity systems based on SAML2 or OpenID Connect Core 1.0, already in production in each Member State (eg: the Italian SPID and CIE id schemes notified eIDAS with *LoA* **High**, see `SPID/CIE OpenID Connect Specifications <https://italia.github.io/spid-cie-oidc-docs/en/>`_).

>     - *(Q)EAA Provider*: It represents the Issuer of (Q)EAAs. It is composed by:

Reviewer: ```suggestion
    - *(Q)EAA Provider*: It represents the Issuer of (Q)EAAs. It is composed with the following components:
```

  * provides to the PID/(Q)EAA Provider PAR endpoint the parameters previously listed above, using the ``request`` parameter (hereafter Request Object) according to :rfc:`9126` Section 3 to prevent Request URI swapping attack.
  * MUST create the ``code_verifier`` with enough entropy random string using the unreserved characters with a minimum length of 43 characters and a maximum length of 128 characters, making it impractical for an attacker to guess its value. The value MUST be generated following the recommendation in Section 4.1 of :rfc:`7636`.
  * signs this request using the private key that is created during the setup phase to obtain the Wallet Attestation. The related public key that is attested by the Wallet Provider is provided within the Wallet Attestation ``cnf`` claim.

>   * MUST the ``OAuth-Client-Attestation`` and  ``OAuth-Client-Attestation-PoP`` parameters according to OAuth 2.0 Attestation-based Client Authentication [`OAUTH-ATTESTATION-CLIENT-AUTH`_], since in this flow the Pushed Authorization Endpoint is a protected endpoint. The first JWT is the Wallet Attestation JWT and the second JWT is the Wallet Attestation Proof of Possession.

Reviewer: ```suggestion
  * MUST use the ``OAuth-Client-Attestation`` and  ``OAuth-Client-Attestation-PoP`` parameters according to OAuth 2.0 Attestation-based Client Authentication [`OAUTH-ATTESTATION-CLIENT-AUTH`_], since in this flow the Pushed Authorization Endpoint is a protected endpoint.
```

>   * MUST create the value of the ``OAuth-Client-Attestation`` and  ``OAuth-Client-Attestation-PoP`` parameter according to OAuth 2.0 Attestation-based Client Authentication [`OAUTH-ATTESTATION-CLIENT-AUTH`_ draft 03], since in this flow the Pushed Authorization Endpoint is a protected endpoint. The first JWT is the Wallet Attestation JWT and the second JWT is the Wallet Attestation Proof of Possession.

Reviewer: ```suggestion
  * MUST create the value of the ``OAuth-Client-Attestation`` and  ``OAuth-Client-Attestation-PoP`` parameter according to OAuth 2.0 Attestation-based Client Authentication [`OAUTH-ATTESTATION-CLIENT-AUTH`_], since in this flow the Pushed Authorization Endpoint is a protected endpoint. The first JWT is the Wallet Attestation JWT and the second JWT is the Wallet Attestation Proof of Possession.
```

>   * MUST create the value of the ``OAuth-Client-Attestation`` and  ``OAuth-Client-Attestation-PoP`` parameter according to OAuth 2.0 Attestation-based Client Authentication [`oauth-attestation-draft <https://datatracker.ietf.org/doc/draft-ietf-oauth-attestation-based-client-auth/>`_], since in this flow the Pushed Authorization Endpoint is a protected endpoint. The first JWT is the Wallet Attestation JWT and the second JWT is the Wallet Attestation Proof of Possession.

Reviewer: please use ref to defined specs and not URLs

Reviewer: this url must be configured here: https://github.com/italia/eudi-wallet-it-docs/blob/versione-corrente/docs/common/standards.rst

and referenced in the text included in this PR

Reviewer 2: done

Reviewer: therefore this url must be replaced with the ref you have defined

>   * MUST create the value of the ``OAuth-Client-Attestation`` and  ``OAuth-Client-Attestation-PoP`` parameter according to OAuth 2.0 Attestation-based Client Authentication [`oauth-attestation-draft <https://datatracker.ietf.org/doc/draft-ietf-oauth-attestation-based-client-auth/>`_], since in this flow the Pushed Authorization Endpoint is a protected endpoint. The first JWT is the Wallet Attestation JWT and the second JWT is the Wallet Attestation Proof of Possession.

Reviewer 2: ```suggestion
  * MUST create the value of the ``OAuth-Client-Attestation`` and  ``OAuth-Client-Attestation-PoP`` parameter according to OAuth 2.0 Attestation-based Client Authentication [`OAUTH-ATTESTATION-CLIENT-AUTH`_ draft 03], since in this flow the Pushed Authorization Endpoint is a protected endpoint. The first JWT is the Wallet Attestation JWT and the second JWT is the Wallet Attestation Proof of Possession.
```

    8. It MUST check that the Request Object is not expired, checking the ``exp`` claim.
    9. It MUST check that the Request Object was issued in a previous time than the value exposed in the ``iat`` claim. It SHOULD reject the request if the ``iat`` claim is far from the current time (:rfc:`9126`) of more than `5` minutes.
    10. It MUST check that the ``jti`` claim in the Request Object has not been used before by the Wallet Instance identified by the ``client_id``. This allows the PID/(Q)EAA Provider to mitigate replay attacks (:rfc:`7519`).

>     11. It MUST validate the ``OAuth-Client-Attestation-PoP`` parameter based on Section 4 of [`OAUTH-ATTESTATION-CLIENT-AUTH`_ draft 03].

Reviewer: ```suggestion
    11. It MUST validate the ``OAuth-Client-Attestation-PoP`` parameter based on Section 4 of [`OAUTH-ATTESTATION-CLIENT-AUTH`_].
```

>     11. It MUST validate the ``OAuth-Client-Attestation-PoP`` parameter based on Section 4 of [`oauth-attestation-draft <https://datatracker.ietf.org/doc/draft-ietf-oauth-attestation-based-client-auth/>`_].

Reviewer 2: ```suggestion
    11. It MUST validate the ``OAuth-Client-Attestation-PoP`` parameter based on Section 4 of [`OAUTH-ATTESTATION-CLIENT-AUTH`_ draft 03].
```

    POST /as/par HTTP/1.1
    Host: eaa-provider.example.org
    Content-Type: application/x-www-form-urlencoded
    OAuth-Client-Attestation: eyJhbGciOiJFUzI1NiIsImtpZCI6IkVVRzBFdlRWaUk1RU5aQXdVQ0lVTWdQQVk4X1VISW5fMkhIWlMxN3RfQzAifQ.eyJpc3MiOiAiaHR0cHM6Ly9jbGllbnQuZXhhbXBsZS5jb20iLCAiYXVkIjogImh0dHBzOi8vYXMuZXhhbXBsZS5jb20iLCAibmJmIjogMTMwMDgxNTc4MCwgImV4cCI6IDEzMDA4MTkzODB9._v3bjJelKI0TNpbc4ysS7yJupwSZzMPQ0ZQ9N5zj8XGQ_T3NN9bghUyVzegR60xokqBnqmMS4iYgPOL7ekEspw

>     OAuth-Client-Attestation-PoP: eyJhbGciOiJFUzI1NiIsImtpZCI6IkVVRzBFdlRWaUk1RU5aQXdVQ0lVTWdQQVk4X1VISW5fMkhIWlMxN3RfQzAifQ.eyJpc3MiOiAiaHR0cHM6Ly9jbGllbnQuZXhhbXBsZS5jb20iLCAiYXVkIjogImh0dHBzOi8vYXMuZXhhbXBsZS5jb20iLCAibmJmIjogMTMwMDgxNTc4MCwgImV4cCI6IDEzMDA4MTkzODB9._v3bjJelKI0TNpbc4ysS7yJupwSZzMPQ0ZQ9N5zj8XGQ_T3NN9bghUyVzegR60xokqBnqmMS4iYgPOL7ekEspw

Reviewer 2: ```suggestion
    OAuth-Client-Attestation-PoP: eyJhbGciOiJFUzI1NiJ9.eyJpc3MiOiIgaHR0cHM6Ly9jbGllbnQuZXhhbXBsZS5jb20iLCJhdWQiOiIgaHR0cHM6Ly9hcy5leGFtcGxlLmNvbSIsImp0aSI6IjVlZmY5YzFiLWVkMGQtNDdlOC1hNTUzLWY3NGRmMWJiZWVkZCIsImlhdCI6MTcyMjI0OTQ0NywiZXhwIjoxNzIyMjQ5NzQ3fQ.aZpx7u7R-W8q7fJh9BEaRf8LM7RQRxAVc-okalAVqxHWqUMh3ehYukMLaCsiDQ33pyS41Y5PEsZ3HXwAXQ3nMg
```

Reviewer: `jwt-client-attestation-pop` typ header is unspecified, therefore what would be the reason it get it here?

Reviewer 2: You are right, I removed typ in the jwt header

Reviewer: therefore the original example that I have changed was fine, do you confirm?

    Host: eaa-provider.example.org
    Content-Type: application/x-www-form-urlencoded
    DPoP: eyJ0eXAiOiJkcG9wK2p3dCIsImFsZyI6IkVTMjU2IiwiandrIjp7Imt0eSI6Ik
    OAuth-Client-Attestation: eyJhbGciOiJFUzI1NiIsImtpZCI6IkVVRzBFdlRWaUk1RU5aQXdVQ0lVTWdQQVk4X1VISW5fMkhIWlMxN3RfQzAifQ.eyJpc3MiOiAiaHR0cHM6Ly9jbGllbnQuZXhhbXBsZS5jb20iLCAiYXVkIjogImh0dHBzOi8vYXMuZXhhbXBsZS5jb20iLCAibmJmIjogMTMwMDgxNTc4MCwgImV4cCI6IDEzMDA4MTkzODB9._v3bjJelKI0TNpbc4ysS7yJupwSZzMPQ0ZQ9N5zj8XGQ_T3NN9bghUyVzegR60xokqBnqmMS4iYgPOL7ekEspw

>     OAuth-Client-Attestation-PoP: eyJhbGciOiJFUzI1NiIsImtpZCI6IkVVRzBFdlRWaUk1RU5aQXdVQ0lVTWdQQVk4X1VISW5fMkhIWlMxN3RfQzAifQ.eyJpc3MiOiAiaHR0cHM6Ly9jbGllbnQuZXhhbXBsZS5jb20iLCAiYXVkIjogImh0dHBzOi8vYXMuZXhhbXBsZS5jb20iLCAibmJmIjogMTMwMDgxNTc4MCwgImV4cCI6IDEzMDA4MTkzODB9._v3bjJelKI0TNpbc4ysS7yJupwSZzMPQ0ZQ9N5zj8XGQ_T3NN9bghUyVzegR60xokqBnqmMS4iYgPOL7ekEspw

Reviewer 2: ```suggestion
    OAuth-Client-Attestation-PoP: eyJhbGciOiJFUzI1NiJ9.eyJpc3MiOiIgaHR0cHM6Ly9jbGllbnQuZXhhbXBsZS5jb20iLCJhdWQiOiIgaHR0cHM6Ly9hcy5leGFtcGxlLmNvbSIsImp0aSI6IjVlZmY5YzFiLWVkMGQtNDdlOC1hNTUzLWY3NGRmMWJiZWVkZCIsImlhdCI6MTcyMjI0OTQ0NywiZXhwIjoxNzIyMjQ5NzQ3fQ.aZpx7u7R-W8q7fJh9BEaRf8LM7RQRxAVc-okalAVqxHWqUMh3ehYukMLaCsiDQ33pyS41Y5PEsZ3HXwAXQ3nMg
```

Reviewer: `jwt-client-attestation-pop` typ header is unspecified, therefore what would be the reason it get it here?

Reviewer 2: fixed

    HTTP/1.1 302 Found
    Location: https://start.wallet.example.org?code=SplxlOBeZQQYbYS6WxSbIA&state=fyZiOL9Lf2CeKuNT2JzxiLRDink0uPcd&iss=https%3A%2F%2Fpid-provider.example.org

> **Steps 12-13 (DPoP Proof for Token Endpoint)**: The Wallet Instance MUST create a new key pair for the DPoP and a fresh DPoP Proof JWT following the instruction provided in this `specific <https://github.com/oauth-wg/draft-ietf-oauth-attestation-based-client-auth/blob/main/draft-ietf-oauth-attestation-based-client-auth.md>`_ for the token request to the PID/(Q)EAA Provider. The DPoP Proof JWT is signed using the private key for DPoP created by Wallet Instance for this scope. DPoP binds the Access Token to a certain Wallet Instance (:rfc:`9449`) and mitigates the misuse of leaked or stolen Access Tokens at the Credential Endpoint.

Reviewer: no, DPoP is not attestation based client auth

**Step 14 (Token Request):** The Wallet Instance sends a token request to the PID/(Q)EAA Provider Token Endpoint with a *DPoP Proof JWT*  and the parameters: ``code``, ``code_verifier``, and OAuth 2.0 Attestation based Client Authentication  (``client_assertion_type=urn:ietf:params:oauth:client-assertion-type:jwt-client-attestation`` and ``client_assertion=WIA~WIA-PoP``). 
The ``client_assertion`` is signed using the private key that is created during the setup phase to obtain the Wallet Attestation. The related public key that is attested by the Wallet Provider is provided within the Wallet Attestation (``cnf`` claim). The PID/(Q)EAA Provider performs the following checks on the Token Request:

   1. It MUST ensure that the Authorization ``code`` is issued to the authenticated Wallet Instance (:rfc:`6749`) and was not replied.
   2. It MUST ensure the Authorization ``code`` is valid and has not been previously used (:rfc:`6749`).
   3. It MUST ensure the ``redirect_uri`` matches the value included in the previous Request Object `OpenID.Core#TokenRequest <https://openid.net/specs/openid-connect-core-1_0.html#TokenRequest>`_.

>    4. It MUST validate the DPoP Proof JWT, according to `OAuth 2.0 Attestation-Based Client Authentication <https://github.com/oauth-wg/draft-ietf-oauth-attestation-based-client-auth/blob/main/draft-ietf-oauth-attestation-based-client-auth.md>`_.

Reviewer: no

    HTTP/1.1 302 Found
    Location: https://start.wallet.example.org?code=SplxlOBeZQQYbYS6WxSbIA&state=fyZiOL9Lf2CeKuNT2JzxiLRDink0uPcd&iss=https%3A%2F%2Fpid-provider.example.org

> **Steps 12-13 (DPoP Proof for Token Endpoint)**: The Wallet Instance MUST create a new key pair for the DPoP and a fresh DPoP Proof JWT following the instruction provided in this Section 4 of (:rfc:`9449`) for the token request to the PID/(Q)EAA Provider. The DPoP Proof JWT is signed using the private key for DPoP created by Wallet Instance for this scope. DPoP binds the Access Token to a certain Wallet Instance (:rfc:`9449`) and mitigates the misuse of leaked or stolen Access Tokens at the Credential Endpoint.

Reviewer: ```suggestion
**Steps 12-13 (DPoP Proof for Token Endpoint)**: The Wallet Instance MUST create a new key pair for the DPoP and a fresh DPoP Proof JWT following the instruction provided in  the Section 4 of (:rfc:`9449`) for the token request to the PID/(Q)EAA Provider. The DPoP Proof JWT is signed using the private key for DPoP created by Wallet Instance for this scope. DPoP binds the Access Token to a certain Wallet Instance (:rfc:`9449`) and mitigates the misuse of leaked or stolen Access Tokens at the Credential Endpoint.
```

    Host: pid-provider.example.org
    Content-Type: application/x-www-form-urlencoded
    DPoP: eyJ0eXAiOiJkcG9wK2p3dCIsImFsZyI6IkVTMjU2IiwiandrIjp7Imt0eSI6Ik
    OAuth-Client-Attestation: eyJhbGciOiAiRVMyNTYiLCJraWQiOiAiMTEifQ.eyJ
        pc3MiOiJodHRwczovL2NsaWVudC5leGFtcGxlLmNvbSIsInN1YiI6Imh0dHBzOi8vY2x
        pZW50LmV4YW1wbGUuY29tIiwibmJmIjoxMzAwODE1NzgwLCJleHAiOjEzMDA4MTkzODA
        sImNuZiI6eyJqd2siOnsia3R5IjoiRUMiLCJ1c2UiOiJzaWciLCJjcnYiOiJQLTI1NiI
        sIngiOiIxOHdITGVJZ1c5d1ZONlZEMVR4Z3BxeTJMc3pZa01mNko4bmpWQWlidmhNIiw
        ieSI6Ii1WNGRTNFVhTE1nUF80Zlk0ajhpcjdjbDFUWGxGZEFnY3g1NW83VGtjU0EifX1
        9.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c
    OAuth-Client-Attestation-PoP: eyJ0eXAiOiJkcG9wK2p3dCIsImFsZyI6IkVTMjU2IiwiandrIjp7Imt0eSI6Ik
        VDIiwieCI6Imw4dEZyaHgtMzR0VjNoUklDUkRZOXpDa0RscEJoRjQyVVFVZldWQVdCR
        nMiLCJ5IjoiOVZFNGpmX09rX282NHpiVFRsY3VOSmFqSG10NnY5VERWclUwQ2R2R1JE
        QSIsImNydiI6IlAtMjU2In19.eyJqdGkiOiItQndDM0VTYzZhY2MybFRjIiwiaHRtIj
        oiUE9TVCIsImh0dSI6Imh0dHBzOi8vc2VydmVyLmV4YW1wbGUuY29tL3Rva2VuIiwia
        WF0IjoxNTYyMjYyNjE2fQ.2-GxA6T8lP4vfrg8v-FdWP0A0zdrj8igiMLvqRMUvwnQg

>         4PtFLbdLXiOSsX0x7NVY-FNyJK70nfbV37xRZT3Lg    

Reviewer: ```suggestion
        4PtFLbdLXiOSsX0x7NVY-FNyJK70nfbV37xRZT3Lg
```

  Content-Type: application/json
  Authorization: DPoP Kz~8mXK1EalYznwH-LC-1fBAo.4Ljp~zsPE_NeO.gxU
  DPoP: eyJ0eXAiOiJkcG9wK2p3dCIsImFsZyI6IkVTMjU2IiwiandrIjp7Imt0eSI6Ik

>   OAuth-Client-Attestation: eyJhbGciOiAiRVMyNTYiLCJraWQiOiAiMTEifQ.eyJ
>         pc3MiOiJodHRwczovL2NsaWVudC5leGFtcGxlLmNvbSIsInN1YiI6Imh0dHBzOi8vY2x
>         pZW50LmV4YW1wbGUuY29tIiwibmJmIjoxMzAwODE1NzgwLCJleHAiOjEzMDA4MTkzODA
>         sImNuZiI6eyJqd2siOnsia3R5IjoiRUMiLCJ1c2UiOiJzaWciLCJjcnYiOiJQLTI1NiI
>         sIngiOiIxOHdITGVJZ1c5d1ZONlZEMVR4Z3BxeTJMc3pZa01mNko4bmpWQWlidmhNIiw
>         ieSI6Ii1WNGRTNFVhTE1nUF80Zlk0ajhpcjdjbDFUWGxGZEFnY3g1NW83VGtjU0EifX1
>         9.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c
>   OAuth-Client-Attestation-PoP: eyJ0eXAiOiJkcG9wK2p3dCIsImFsZyI6IkVTMjU2IiwiandrIjp7Imt0eSI6Ik

Reviewer 2: ```suggestion
```

Reviewer 2: The Credential Endpoint is not protected by Client Authentication.

Pushed Authorization Request (PAR) Request
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The requests to the PID/(Q)EAA authorization endpoint MUST use the HTTP POST method with the parameters in the message body encoded in ``application/x-www-form-urlencoded`` format. The Pushed Authorization Endpoint is protected with OAuth 2.0 Attestation-based Client Authentication [`OAUTH-ATTESTATION-CLIENT-AUTH`_ draft 03] and the following parameters MUST be provided:

Reviewer: ```suggestion
The requests to the PID/(Q)EAA authorization endpoint MUST use the HTTP POST method with the parameters in the message body encoded in ``application/x-www-form-urlencoded`` format. The Pushed Authorization Endpoint is protected with OAuth 2.0 Attestation-based Client Authentication [`OAUTH-ATTESTATION-CLIENT-AUTH`_] and the following parameters MUST be provided:
```

> The requests to the PID/(Q)EAA authorization endpoint MUST use the HTTP POST method with the parameters in the message body encoded in ``application/x-www-form-urlencoded`` format. The Pushed Authorization Endpoint is protected with OAuth 2.0 Attestation-based Client Authentication [`oauth-attestation-draft <https://datatracker.ietf.org/doc/draft-ietf-oauth-attestation-based-client-auth/>`_] and the following parameters MUST be provided:

Reviewer 2: ```suggestion
The requests to the PID/(Q)EAA authorization endpoint MUST use the HTTP POST method with the parameters in the message body encoded in ``application/x-www-form-urlencoded`` format. The Pushed Authorization Endpoint is protected with OAuth 2.0 Attestation-based Client Authentication [`OAUTH-ATTESTATION-CLIENT-AUTH`_ draft 03] and the following parameters MUST be provided:
```

    * - **request**
      - It MUST be a signed JWT. The private key corresponding to the public one in the ``cnf`` parameter inside the Wallet Attestation MUST be used for signing the Request Object.
      - `OpenID Connect Core. Section 6 <https://openid.net/specs/openid-connect-core-1_0.html#JWTRequests>`_
    * - **OAuth-Client-Attestation**
      - It MUST be set to a value containing the Wallet Attestation JWT.

>       - `OAUTH-ATTESTATION-CLIENT-AUTH`_ draft 03.

Reviewer: ```suggestion
      - `OAUTH-ATTESTATION-CLIENT-AUTH`_.
```

>     * - **OAuth-Client-Attestation**
>       - It MUST be set to a value containing the Wallet Attestation JWT.
>       - `OAUTH-ATTESTATION-CLIENT-AUTH`_ draft 03.
>     * - **OAuth-Client-Attestation-PoP**
>       - It MUST be set to a value containing the Wallet Attestation JWT Proof of Possession.
>       - `OAUTH-ATTESTATION-CLIENT-AUTH`_ draft 03.

Reviewer 2: ```suggestion
```

Reviewer 2: These parameters must be included in the HTTP header request and not in the payload of the request object.

Reviewer: they must be moved in a separate table defining the http request header parameters

as @m-basili mentioned, in the current PR this table is only related to the http post body parameters and not http request header parameters

>       - `OAUTH-ATTESTATION-CLIENT-AUTH`_ draft 03.

Reviewer: ```suggestion
      - `OAUTH-ATTESTATION-CLIENT-AUTH`_.
```

>     * - **OAuth-Client-Attestation**
>       - It MUST be set to a value containing the Wallet Attestation JWT.
>       - `oauth-attestation-draft <https://datatracker.ietf.org/doc/draft-ietf-oauth-attestation-based-client-auth/>`_.
>     * - **OAuth-Client-Attestation-PoP**
>       - It MUST be set to a value containing the Wallet Attestation JWT Proof of Possession.
>       - `oauth-attestation-draft <https://datatracker.ietf.org/doc/draft-ietf-oauth-attestation-based-client-auth/>`_.

Reviewer 2: ```suggestion
    * - **OAuth-Client-Attestation**
      - It MUST be set to a value containing the Wallet Attestation JWT.
      - `OAUTH-ATTESTATION-CLIENT-AUTH`_ draft 03.
    * - **OAuth-Client-Attestation-PoP**
      - It MUST be set to a value containing the Wallet Attestation JWT Proof of Possession.
      - `OAUTH-ATTESTATION-CLIENT-AUTH`_ draft 03.
```

    * - **redirect_uri**
      -  Redirection URI to which the response is intended to be sent. It MUST be an universal or app link registered with the local operating system, so this latter will provide the response to the Wallet Instance.
      - See `OpenID.Core#AuthRequest <https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest>`_.

>     * - **OAuth-Client-Attestation**
>       - It MUST be set as in the :ref:`Table of the HTTP parameters <table_http_request_claim>`.
>       - See :ref:`Table of the HTTP parameters <table_http_request_claim>`.
>     * - **OAuth-Client-Attestation-PoP**
>       - It MUST be set as in the :ref:`Table of the HTTP parameters <table_http_request_claim>`.
>       - See :ref:`Table of the HTTP parameters <table_http_request_claim>`.

Reviewer 2: ```suggestion
```

Reviewer: see my previous comment, it confirms @m-basili 's review and asks a specific table about the http request header parameters to be included within the request

    * - **alg**
      - A digital signature algorithm identifier such as per IANA "JSON Web Signature and Encryption Algorithms" registry. It MUST be one of the supported algorithms listed in the Section `Cryptographic Algorithms <algorithms.html>`_ and MUST NOT be set to ``none`` or any symmetric algorithm (MAC) identifier.
      - :rfc:`7516#section-4.1.1`.

>     * - **typ**
>       -  It MUST be set to ``jwt-client-attestation-pop``
>       -  Currently under discussion in [`oauth-attestation-draft <https://vcstuff.github.io/draft-ietf-oauth-attestation-based-client-auth/draft-ietf-oauth-attestation-based-client-auth.html>`_].

Reviewer 2: ```suggestion
```

Reviewer 2: @peppelinux, check please.

      - Unique identifier of the JWT that, together with the value contained in the ``iss`` claim,  prevents the reuse of the JWT (replay attack). Since the `jti` value alone is not collision resistant, it MUST be identified uniquely together with its issuer.
      - [:rfc:`7519`].

> The JOSE header of the Wallet Attestation proof of possession MUST contains:

Reviewer: ```suggestion
The JOSE header of the Wallet Attestation proof of possession MUST contain:
```

      - **Description**
      - **Reference**
    * - **alg**

>       - OPTIONAL. A digital signature algorithm identifier such as per IANA "JSON Web Signature and Encryption Algorithms" registry. It MUST be one of the supported algorithms listed in the Section `Cryptographic Algorithms <algorithms.html>`_ and MUST NOT be set to ``none`` or any symmetric algorithm (MAC) identifier.

Reviewer: ```suggestion
      - REQUIRED. A digital signature algorithm identifier such as per IANA "JSON Web Signature and Encryption Algorithms" registry. It MUST be one of the supported algorithms listed in the Section `Cryptographic Algorithms <algorithms.html>`_ and MUST NOT be set to ``none`` or any symmetric algorithm (MAC) identifier.
```

@cmarco0 why did you configure these as optional?

      - :rfc:`7516#section-4.1.1`.
    * - **kid**
      -  OPTIONAL. Unique identifier of the ``jwk`` inside the ``cnf`` claim of Wallet Attestation as base64url-encoded JWK Thumbprint value.

>       - :rfc:`7638#section_3`.

Reviewer: @fmarino-ipzs it seems to me that kid to be references/contained in cnf.jwk doesn't have sense

may we remove this?

    * - **typ**

>       -  OPTIONAL. It MUST be set to ``jwt-client-attestation-pop``

Reviewer: ```suggestion
      -  REQUIRED. It MUST be set to ``jwt-client-attestation-pop``
```

      -  Currently under discussion in [`oauth-attestation-draft <https://vcstuff.github.io/draft-ietf-oauth-attestation-based-client-auth/draft-ietf-oauth-attestation-based-client-auth.html>`_].

> The body of the Wallet Attestation proof of possession JWT contains:

Reviewer: @cmarco0 please don't change the requirements, the scope of this PR is to only change the HTTP Request POST parameters related to the attestation based client auth to the HTTP Request header parameters as defined in the latest draft.

      -  It MUST be set to ``jwt-client-attestation-pop``
      -  Currently under discussion in [`oauth-attestation-draft <https://vcstuff.github.io/draft-ietf-oauth-attestation-based-client-auth/draft-ietf-oauth-attestation-based-client-auth.html>`_].

> The body of the Wallet Attestation proof of possession JWT MUST contains:

Reviewer: ```suggestion
The body of the Wallet Attestation proof of possession JWT MUST contain:
```

--------------

The token endpoint is used by the Wallet Instance to obtain an Access Token by presenting an authorization grant, as

> defined in :rfc:`6749`. The Token Endpoint is a protected endpoint with a client authentication based on the model defined in OAuth 2.0 Attestation-based Client Authentication [`OAUTH-ATTESTATION-CLIENT-AUTH`_ draft 03].

Reviewer: ```suggestion
defined in :rfc:`6749`. The Token Endpoint is a protected endpoint with a client authentication based on the model defined in OAuth 2.0 Attestation-based Client Authentication [`OAUTH-ATTESTATION-CLIENT-AUTH`_ ].
```

I suggest to keep draft version out from the text and only defining the draft url in the specs href pointed by the ref

> defined in :rfc:`6749`. The Token Endpoint is a protected endpoint with a client authentication based on the model defined in OAuth 2.0 Attestation-based Client Authentication [`oauth-attestation-draft <https://datatracker.ietf.org/doc/draft-ietf-oauth-attestation-based-client-auth/>`_].

Reviewer 2: ```suggestion
defined in :rfc:`6749`. The Token Endpoint is a protected endpoint with a client authentication based on the model defined in OAuth 2.0 Attestation-based Client Authentication [`OAUTH-ATTESTATION-CLIENT-AUTH`_ draft 03].
```

.. _sec_token_request:

Token Request
^^^^^^^^^^^^^^^

> The request to the PID/(Q)EAA Token endpoint MUST be an HTTP request with method POST, with the body message encoded in ``application/x-www-form-urlencoded`` format. The Wallet Instance sends the Token endpoint request with ``OAuth-Client-Attestation`` and ``OAuth-Client-Attestation-PoP``.

Reviewer 2: ```suggestion
The request to the PID/(Q)EAA Token endpoint MUST be an HTTP request with method POST, with the body message encoded in ``application/x-www-form-urlencoded`` format. The Wallet Instance sends the Token endpoint request with ``OAuth-Client-Attestation`` and ``OAuth-Client-Attestation-PoP`` as header parameters according to `OAUTH-ATTESTATION-CLIENT-AUTH`_.
```

Reviewer: remove draft 03

    * - **code_verifier**
      - Verification code of the **code_challenge**.
      - `Proof Key for Code Exchange by OAuth Public Clients <https://datatracker.ietf.org/doc/html/rfc7636>`_.

>     * - **OAuth-Client-Attestation**
>       - It MUST be set to a value containing the Wallet Attestation JWT.
>       - `OAUTH-ATTESTATION-CLIENT-AUTH`_ draft 03.
>     * - **OAuth-Client-Attestation-PoP**
>       - It MUST be set to a value containing the Wallet Attestation JWT Proof of Possession JWT. The Wallet Attestation Proof of Possession MUST contain the claims as defined in :ref:`Table of the JWT Wallet Attestation PoP <table_jwt_pop>`, Section :ref:`Pushed Authorization Request (PAR) Response<sec_par>`.
>       - `OAUTH-ATTESTATION-CLIENT-AUTH`_ draft 03.

Reviewer 2: ```suggestion
```

>     * - **OAuth-Client-Attestation**
>       - It MUST be set to a value containing the Wallet Attestation JWT.
>       - `oauth-attestation-draft <https://datatracker.ietf.org/doc/draft-ietf-oauth-attestation-based-client-auth/>`_.
>     * - **OAuth-Client-Attestation-PoP**
>       - It MUST be set to a value containing the Wallet Attestation JWT Proof of Possession JWT. The Wallet Attestation Proof of Possession MUST contain the claims as defined in :ref:`Table of the JWT Wallet Attestation PoP <table_jwt_pop>`, Section :ref:`Pushed Authorization Request (PAR) Response<sec_par>`.
>       - `oauth-attestation-draft <https://datatracker.ietf.org/doc/draft-ietf-oauth-attestation-based-client-auth/>`_.

Reviewer 2: ```suggestion
    * - **OAuth-Client-Attestation**
      - It MUST be set to a value containing the Wallet Attestation JWT.
      - `OAUTH-ATTESTATION-CLIENT-AUTH`_ draft 03.
    * - **OAuth-Client-Attestation-PoP**
      - It MUST be set to a value containing the Wallet Attestation JWT Proof of Possession JWT. The Wallet Attestation Proof of Possession MUST contain the claims as defined in :ref:`Table of the JWT Wallet Attestation PoP <table_jwt_pop>`, Section :ref:`Pushed Authorization Request (PAR) Response<sec_par>`.
      - `OAUTH-ATTESTATION-CLIENT-AUTH`_ draft 03.
```
